### ISSUE TITLE
Generating larger Sets is slow

### ISSUE THREAD

**Issue by martinei** [jlink/jqwik#203]
## Testing Problem

I observed the following:
```
    @Test
    void fastSet() { // takes ~200ms
        var list = Arbitraries.defaultFor(String.class).list().ofSize(15000).sample();
        var set = new HashSet(list);
        System.out.println (set.size()); // Produces 13K to 14K strings
    }
    @Test
    
    void slowSet() { // takes ~ 4-5 seconds
        Arbitraries.defaultFor(String.class).set().ofSize(10000).sample();
    }
```
It looks like first a number of strings is generated and the resulting list is then checked for uniqueness. This is repeated until a unique set is found.

**Comment by jlink** [comment]
Thanks for noticing. Actually the list is checked on each insert for uniqueness. But still, there's probably some optimization potential for sets.

Up to now I didn't feel the need for optimization myself. Can you specify how large your typical set is?

**Comment by martinei** [comment]
Currently I have try to generate a set around 15K Items. My actual code uses a domain class class where equals / hashcode are a bit more complex (and thus slower) than with strings.

I can workaround by getting a Stream<T> and collecting into a HashSet until my wanted size is reached. Or just getting a larger list and hope that the remaining set after uniqueness is big enough.

**Comment by jlink** [comment]
> Or just getting a larger list and hope that the remaining set after uniqueness is big enough.

Hoping is the problem of this approach ;-)

I'll think about how to optimize while keeping all the other constraints in place.

**Comment by jlink** [comment]
OK, I spent the day (well a few hours) trying to optimize set generation. So far, I haven't found a way to speed it up in a measurable way without breaking some of the features (eg. deterministic size, other uniqueness constraints).

That's why my current recommendation for you is to use 
`domainArbitrary.list().ofSize(10000).map(HashSet::new)`.

What you lose compared to native set generation is the size guarantee (in an extreme case the set may have a single element), advanced uniqueness features and shrinking won't work as well either. 

If your domain class generator is producing duplicates only rarely and if you switch off edge case generation for it, you may want to get rid of edge cases and filter out smaller sets like that:
`domainArbitrary.withoutEdgeCases().list().ofSize(10000).map(HashSet::new).filter(s -> s.size() == 10000)`

Let me know if that works out for you.

**Comment by martinei** [comment]
Thanks for looking into this. I am sorry that I took so much of your time.
Your recommendation  give much faster results.
In addition I finally figured that the total number of possible values I can generate from was restricted to much:
At the heart of my domain Object is a Version Number like:  "<major>.<minor>.<patch>-<prerelease>+<metadata>"
I had restricted the individual parts, and thus the total number of possible values, to much.
I now use bigger Values, like:

```
        private Arbitrary<Integer> major = Arbitraries.integers().between(1, 500).withoutEdgeCases();
        private Arbitrary<Integer> minor = Arbitraries.integers().between(0, 500).withoutEdgeCases();
        private Arbitrary<Integer> patch = Arbitraries.integers().between(0, 500).withoutEdgeCases();
```
And combine them to a Version Number, giving a total of ~ 125Million Combinations.  Now Sets of size 10000 are no problem any longer.
